File: JRuby/k-nucleotide/knucleotide.rb
# The Computer Language Benchmarks Game
# http://benchmarksgame.alioth.debian.org
#
# contributed by jose fco. gonzalez
# modified by Sokolov Yura
# Parallelism by Rick Branson
# converted to threads on ruby Christian Stone

$seq = String.new

def frequency(seq, length)
  frequencies = Hash.new(0)
  ns          = seq.length + 1 - length
  
  for i in (0 ... ns)
    frequencies[seq[i, length]] += 1
  end
  
  [ns, frequencies]
end

def sort_by_freq(seq, length)
  ret       = ""
  n, table  = frequency(seq, length)

  table.sort{|a,b| b[1] <=> a[1]}.each do |v|
      ret += "%s %.3f\n" % [v[0].upcase,((v[1]*100).to_f/n)]
  end
  
  ret += "\n"
end

def find_seq(seq, s)
  n, table = frequency(seq, s.length)
  "#{table[s].to_s}\t#{s.upcase}\n"
end

line = STDIN.gets while line !~ /^>THREE/
line = STDIN.gets
while (line !~ /^>/) & line do
    $seq << line.chomp
    line = STDIN.gets
end

class Worker
  def initialize(&block)
    @t = Thread.new do
      Thread.current[:result] = yield
    end
  end
  
  def result
    @t.join
    @t[:result]
  end
end

FREQS   = [1, 2]
NUCLEOS = %w(ggt ggta ggtatt ggtattttaatt ggtattttaatttatagt)

workers =   FREQS.map   { |i| Worker.new { sort_by_freq($seq, i) } }
workers +=  NUCLEOS.map { |s| Worker.new { find_seq($seq, s) } }
  
results = workers.map { |w| w.result }
print results.join

File: experiments/JavaScript as TypeScript/k-nucleotide/knucleotide.ts
/* The Computer Language Benchmarks Game
   http://benchmarksgame.alioth.debian.org/

   Contributed by Jesse Millikan
   Modified by Matt Baker
   Ported, modified, and parallelized by Roman Pletnev
*/

'use strict';

var rd = require('readline'), cp = require('child_process');

function RefNum(num){ this.num = num; }
RefNum.prototype.toString = function() { return this.num.toString(); }

function frequency(seq, length){
    var freq = new Map(), n = seq.length-length+1, key, cur, i = 0;
    for(; i<n; ++i){
        key = seq.substr(i, length);
        cur = freq.get(key);
        cur === undefined ? freq.set(key, new RefNum(1)) : ++cur.num;
    }
    return freq;
}

function sort(seq, length){
    var f = frequency(seq, length), keys = Array.from(f.keys()),
        n = seq.length-length+1, res = '';
    keys.sort((a, b)=>f.get(b)-f.get(a));
    for (var key of keys) res +=
        key.toUpperCase()+' '+(f.get(key)*100/n).toFixed(3)+'\n';
    res += '\n';
    return res;
}

function find(seq, s){
    var f = frequency(seq, s.length);
    return (f.get(s) || 0)+"\t"+s.toUpperCase()+'\n';
}

function master() {
    var workers = [];
    for (var i=1; i<5;++i) workers.push(
        cp.fork(__filename, [], {silent: true, env: {workerId: i}}));
    for (var w of workers) process.stdin.pipe(w.stdin);
    var jobs = workers.length, results = [];
    var messageHandler = function(i){
        return function(message){
            results[i] = message;
            if (!(--jobs)) {
                process.stdout.write(results.join(''));
                process.exit(0);
            }
        };
    };
    for (var i=0; i<workers.length; ++i)
        workers[i].on('message', messageHandler(i));
}

function worker(){
    var seq = '', reading = false;
    var lineHandler = function(line){
        if (reading) {
            if (line[0]!=='>') seq += line;
        } else reading = line.substr(0, 6)==='>THREE';
    };
    rd.createInterface(process.stdin, process.stdout)
        .on('line', lineHandler).on('close', function() {
            var res = '';
            switch (process.env.workerId) {
                case '1':
                    res += sort(seq, 1);
                    res += sort(seq, 2);
                    res += find(seq, "ggt");
                    break;
                case '2':
                    res += find(seq, "ggta");
                    res += find(seq, "ggtatt");
                    break;
                case '3':
                    res += find(seq, "ggtattttaatt");
                    break;
                case '4':
                    res += find(seq, "ggtattttaatttatagt");
                    break;
            }
            process.send(res);
            process.exit();
        });
}

process.env.workerId ? worker() : master();